src/smc-webapp/project_store.coffee

```diff
@@ -829,18 +829,30 @@ class ProjectActions extends Actions
             zip_args : undefined
             path     : undefined   # default to root of project
             id       : undefined
-        id = opts.id ? misc.uuid()
-        @set_activity(id:id, status:"Creating #{opts.dest} from #{opts.src.length} #{misc.plural(opts.src.length, 'file')}")
+            cb       : undefined
         args = (opts.zip_args ? []).concat(['-rq'], [opts.dest], opts.src)
-        webapp_client.exec
-            project_id      : @project_id
-            command         : 'zip'
-            args            : args
-            timeout         : 50
-            network_timeout : 60
-            err_on_exit     : true    # this should fail if exit_code != 0
-            path            : opts.path
-            cb              : @_finish_exec(id)
+        if (opts.cb?)
```

Comment:
Don’t copy/paste the same code twice like this; doing that is not pretty and hard to mainting.

The entire thing you’ve done here is equivalent to just doing it once and making the input for cb be

 `opts.cb ? @_finish_exec(id)`

which means `opts.cb` if it is defined, and `@_finish_exec(id)` if it isn’t.

Reply:
I choose to write it like that for [lines 845 & 846](https://github.com/sagemathinc/cocalc/pull/2341/files#diff-e6453b90cf71dc76e152191a2985bafdR845), which would have no clean up if the callback was already given. What would be the best way to handle that?

Reply:
With an if statement.  There’s absolutely no possible reason ever — except for one-off code (so not part of a production codebase) — to write copy that has an exact copy of itself later on.